runtime/vm/monhelpers.c

```diff
@@ -150,6 +107,41 @@ objectMonitorExit(J9VMThread* vmStruct, j9object_t object)
 				}
 #endif /* JAVA_SPEC_VERSION >= 24 */
 			}
+		} else {
+			j9objectmonitor_t newLock = 0;
+			I_32 casSuccess = FALSE;
+			if (0x00 == count) {
+				/* just release the flatlock */
+			} else if ((count & OBJECT_HEADER_LOCK_LEARNING) && (count & OBJECT_HEADER_LOCK_LEARNING_RECURSION_MASK)) {
+				/* Learning state case */
+				newLock = lock - OBJECT_HEADER_LOCK_LEARNING_FIRST_RECURSION_BIT;
+			} else if (count & OBJECT_HEADER_LOCK_LEARNING) {
+				/* Lock is in Learning state but unowned (if it were owned it would have been caught by the first Learning state check) */
+				goto done;
+			} else if (count >= OBJECT_HEADER_LOCK_FIRST_RECURSION_BIT) {
+				/* just decrement the flatlock recursion count */
+				newLock = lock - OBJECT_HEADER_LOCK_FIRST_RECURSION_BIT;
+#ifdef J9VM_THR_LOCK_RESERVATION
+			} else if (count & OBJECT_HEADER_LOCK_RESERVED) {
+				/* lock is reserved but unowned (if it were owned the count would be >= OBJECT_HEADER_LOCK_FIRST_RECURSION_BIT) */
+				Trc_VM_objectMonitorExit_Exit_ReservedButUnownedFlatLock(vmStruct, lock, object);
+				goto done;
+#endif
+			}
+
+			monitorExitWriteBarrier();
+			if (J9VMTHREAD_COMPRESS_OBJECT_REFERENCES(vmStruct)) {
+				casSuccess = (lock == compareAndSwapU32((uint32_t*)lockEA, (uint32_t)lock, (uint32_t)newLock));
+			} else {
+				casSuccess = (lock == compareAndSwapUDATA((uintptr_t*)lockEA, (uintptr_t)lock, (uintptr_t)newLock));
+			}
+
+			if (!casSuccess) {
+				/*
+				 * Another thread has attempted to get the lock and may have update the FLC flag.
+				 */
```

Comment:
a multi-line comment is not needed for a single line
```suggestion
				/* Another thread has attempted to get the lock and may have update the FLC flag. */
```